### ISSUE TITLE
Kubelet extra args not getting set on Windows hosts

### ISSUE THREAD

**Issue by dnoland1** [rancher/rancher#22818]
**What kind of request is this (question/bug/enhancement/feature request):**
bug

**Steps to reproduce (least amount of steps as possible):**
When creating a cluster (or edit a cluster), edit as YAML and add:
```yaml
  kubelet:
    extra_args:
      feature-gates: "HyperVContainer=true,WindowsGMSA=true"
```

**Result:**
Linux kubelets have the feature gates set:
```
kubelet --v=2 --pod-infra-container-image=rancher/pause:3.1 --anonymous-auth=false --enforce-node-allocatable= --make-iptables-util-chains=true --client-ca-file=/etc/kubernetes/ssl/kube-ca.pem --hostname-override=fmscaasprel103 --read-only-port=0 --address=0.0.0.0 --resolv-conf=/etc/resolv.conf --kubeconfig=/etc/kubernetes/ssl/kubecfg-kube-node.yaml --root-dir=/var/lib/kubelet --network-plugin=cni --cluster-dns=10.43.0.10 --authentication-token-webhook=true --cloud-provider= --fail-swap-on=false --cni-conf-dir=/etc/cni/net.d --event-qps=0 --streaming-connection-idle-timeout=30m --allow-privileged=true --cgroups-per-qos=True --volume-plugin-dir=/var/lib/kubelet/volumeplugins --feature-gates=WindowsGMSA=true --cni-bin-dir=/opt/cni/bin --cluster-domain=cluster.local --tls-cipher-suites=TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305 --cgroup-driver=cgroupfs
```
But Windows kubelets do not:
```
kubelet.exe              "C:\etc\kubernetes\bin\kubelet.exe" "--network-plugin=cni" "--cni-bin-dir=C:\etc\cni\bin" "--cni-conf-dir=C:\etc\cni\conf" "--logtostderr=false" "--alsologtostderr=true" "--log-file=C:\var\log\kubelet.log"  "--root-dir=c:/var/lib/kubelet" "--pod-infra-container-image=rancher/kubelet-pause:v0.1.2" "--resolv-conf=" "--register-with-taints=beta.kubernetes.io/os=windows:PreferNoSchedule" "--address=0.0.0.0" "--volume-plugin-dir=c:/var/lib/kubelet/volumeplugins" "--authentication-token-webhook=true" "--kubeconfig=c:/etc/kubernetes/ssl/kubecfg-kube-node.yaml" "--cluster-domain=cluster.local" "--cluster-dns=10.43.0.10" "--cgroups-per-qos=false" "--allow-privileged=true" "--read-only-port=0" "--tls-cipher-suites=TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305" "--client-ca-file=c:/etc/kubernetes/ssl/kube-ca.pem" "--cloud-provider=" "--image-pull-progress-deadline=20m" "--node-ip=10.0.0.3" "--cloud-config=" "--feature-gates=HyperVContainer=true" "--enforce-node-allocatable=" "--v=2" "--anonymous-auth=false"  2960
```

**Other details that may be helpful:**
The two features gates mentioned above (HyperVContainer and WindowsGMSA) are now enabled by default, but appears the ability to pass extra args to Windows hosts is not available.

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI):
v2.3.0-alpha5

- Installation option (single install/HA):
n/a

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported):
Custom
- Machine type (cloud/VM/metal) and specifications (CPU/memory):
n/a
- Kubernetes version (use `kubectl version`):
n/a
- Docker version (use `docker version`):
n/a

**Comment by loganhz** [comment]
Maybe we need to make it more generic to support https://github.com/rancher/rancher/issues/23075 as well

**Comment by kinarashah** [comment]
Unable to reproduce on rancher/rancher:v2.3.0, 

Added a custom windows cluster from Rancher UI, following `https://rancher.com/docs/rancher/v2.x/en/cluster-provisioning/rke-clusters/windows-clusters/` 
and added an extra feature-gate (edit as YAML) to the default feature gates for Windows. 

```
    kubelet:
      extra_args:
        feature-gates: 'HyperVContainer=true,WindowsGMSA=true,WatchBookmark=false'
``` 

Confirmed that both linux and windows nodes get the same feature gates. 

Linux node, 
```
echo kubelet --hostname-override=ip-172-31-27-38 --fail-swap-on=false --event-qps=0 --cni-bin-dir=/opt/cni/bin --make-iptables-util-chains=true --pod-infra-container-image=rancher/pause:3.1 --register-with-taints=node-role.kubernetes.io/controlplane=true:NoSchedule --authentication-token-webhook=true --read-only-port=0 --resolv-conf=/etc/resolv.conf --network-plugin=cni --anonymous-auth=false --volume-plugin-dir=/var/lib/kubelet/volumeplugins --client-ca-file=/etc/kubernetes/ssl/kube-ca.pem --root-dir=/var/lib/kubelet --cgroups-per-qos=True --v=2 --authorization-mode=Webhook --cluster-domain=cluster.local --kubeconfig=/etc/kubernetes/ssl/kubecfg-kube-node.yaml --tls-cipher-suites=TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305 --feature-gates=HyperVContainer=true,WindowsGMSA=true,WatchBookmark=false --cloud-provider= --address=0.0.0.0 --streaming-connection-idle-timeout=30m --cni-conf-dir=/etc/cni/net.d --cluster-dns=10.43.0.10
``` 

Windows node, 
```
CommandLine=c:\etc\kubernetes\bin\rancher-wins-kubelet.exe --address=0.0.0.0 --anonymous-auth=false --authentication-token-webhook=true --authorization-mode=Webhook --cert-dir=c:/var/lib/kubelet/pki --cgroups-per-qos=false --client-ca-file=c:/etc/kubernetes/ssl/kube-ca.pem --cloud-provider= --cluster-dns=10.43.0.10 --cluster-domain=cluster.local --cni-bin-dir=c:/opt/cni/bin --cni-conf-dir=c:/etc/cni/net.d --enforce-node-allocatable= --event-qps=0 --fail-swap-on=false --feature-gates=HyperVContainer=true,WindowsGMSA=true,WatchBookmark=false --hostname-override=ec2amaz-deemcuh --image-pull-progress-deadline=30m --kube-reserved=cpu=500m,memory=500Mi,ephemeral-storage=1Gi --kubeconfig=c:/etc/kubernetes/ssl/kubecfg-kube-node.yaml --make-iptables-util-chains=true --network-plugin=cni --node-ip=172.31.28.255 --pod-infra-container-image=rancher/kubelet-pause:v0.1.3 --read-only-port=0 --resolv-conf= --root-dir=c:/var/lib/kubelet --streaming-connection-idle-timeout=30m --system-reserved=cpu=1000m,memory=2Gi,ephemeral-storage=2Gi --tls-cipher-suites=TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305 --v=2 --volume-plugin-dir=c:/var/lib/kubelet/volumeplugins
```

**Comment by sowmyav27** [comment]
**Verified on `2.3.2-rc3`**
**Steps:**

- Create a custom windows cluster - k8s 1.15.
- Linux node - all roles, windows node - worker - 1903 v
- Edit yaml of cluster and enter the below arguments

```
    kubelet:
      extra_args:
        feature-gates: 'HyperVContainer=false,WindowsGMSA=true,WatchBookmark=false'
```
- Both linux and windows nodes get the same feature gates.

**Linux node**
```
"Args": [
            "kubelet",
            "--v=2",
            "--feature-gates=HyperVContainer=false,WindowsGMSA=true,WatchBookmark=false",
            "--cni-conf-dir=/etc/cni/net.d",
            "--address=0.0.0.0",
```

**Windows node:**
```
"Args": [
            "-NoLogo",
            "-NonInteractive",
            "-File",
            "c:/usr/bin/entrypoint.ps1",
            "kubelet",
            "--cni-conf-dir=c:/etc/cni/net.d",
            "--read-only-port=0",
            "--system-reserved=cpu=1000m,memory=2Gi,ephemeral-storage=2Gi",
            "--event-qps=0",
            "--kubeconfig=c:/etc/kubernetes/ssl/kubecfg-kube-node.yaml",
            "--make-iptables-util-chains=true",
            "--cni-bin-dir=c:/opt/cni/bin",
            "--streaming-connection-idle-timeout=30m",
            "--enforce-node-allocatable=",
            "--cluster-dns=x.x.xx",
            "--fail-swap-on=false",
            "--hostname-override=ec2amaz-j67aa1i",
            "--v=2",
            "--authorization-mode=Webhook",
            "--volume-plugin-dir=c:/var/lib/kubelet/volumeplugins",
            "--kube-reserved=cpu=500m,memory=500Mi,ephemeral-storage=1Gi",
            "--image-pull-progress-deadline=30m",
            "--cluster-domain=cluster.local",
            "--authentication-token-webhook=true",
            "--tls-cipher-suites=TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305",
            "--cert-dir=c:/var/lib/kubelet/pki",
            "--pod-infra-container-image=rancher/kubelet-pause:v0.1.3",
            "--node-ip=x.x.x.x",
            "--anonymous-auth=false",
            "--resolv-conf=",
            "--feature-gates=HyperVContainer=false,WindowsGMSA=true,WatchBookmark=false",
```

**Comment by thxCode** [comment]
@sowmyav27 , the above verification could not detect the real problem: **Windows needs a different parameter list against Linux**. The below content shows those unsupported arguments on the Windows side:
- kubelet:
https://github.com/rancher/kontainer-driver-metadata/blob/096e86e3ec17c463deb021a6a5c92f0cdd54b1e9/rke/k8s_windows_service_options.go#L22-L26
- kube-proxy:
https://github.com/rancher/kontainer-driver-metadata/blob/096e86e3ec17c463deb021a6a5c92f0cdd54b1e9/rke/k8s_windows_service_options.go#L65-L70

I give an example to show what happens:
1. Set up a cluster to indicate the `cgroups-per-qos = true`, Linux could apply this, but Windows could not:
    ![屏幕快照 2019-10-30 10 07 05](https://user-images.githubusercontent.com/3394724/67825610-d4ca9d00-fb04-11e9-9d4f-7a9088d2f65f.png)
1. Go ahead to finish the Windows worker adding, you would find the cluster cannot be brought up forever:
    ![image](https://user-images.githubusercontent.com/3394724/67825674-122f2a80-fb05-11e9-9d55-00ba69191991.png)
1. Take a look the kubelet container inside Windows, you could find a FATAL error:
    ![image](https://user-images.githubusercontent.com/3394724/67825727-3854ca80-fb05-11e9-950d-8100878fa057.png)

In short, the component arguments of Windows could not simply overwrite with our existing method. We need a new key (like `kubelet_windows`, `kube-proxy_windows`) or rule (like xxx`_windows` ) to distinguish Windows from Linux arguments.

cc @kinarashah

**Comment by loganhz** [comment]
The original issue has been fixed in 2.3.0. The potential issue which Frank mentioned should be fixed in https://github.com/rancher/rancher/issues/23075
